codeintel: add support to gitserver to download artifacts from maven repositories

shrouxm opened the pull request:
@unknwon still need to test everything, but wanted to ask for a review before i dive into that just to make sure the direction i'm going seems reasonable to you.

some things i'm punting on:
- i'm taking advantage of the logic we use for adding github/gitlab repos to dotcom to solve the problem of not pulling down all of maven at once. this logic was written as a special case, but is starting to look like it should get a proper abstraction. i didn't want to dive into a refactor when i'm so unfamiliar with the gitserver code though
- for VCSSyncer functions which return a command, i'm just allowing them to return nil to indicate that that functionality isn't supported for that syncer

@olafurpg the files vcs_syncer.go and coursier.go contain some logic that still needs to be filled in that you're better equipped to tackle than i. let me know if it seems manageable for you to dive straight into that or if you wanna pair/do some more knowledge transfer beforehand.

i'm going to leave this PR until i get a :+1: that the general direction is good and start working on a PR for #20835 since they're pretty much orthogonal

cc @efritz for context, don't have to review if you don't wanna though

closes #20834

unknwon commented on cmd/repo-updater/shared/main.go:
```diff
@@ -215,6 +215,8 @@ func Main(enterpriseInit EnterpriseInit) {
 					server.GitLabDotComSource, err = repos.NewGitLabSource(e, cf)
 				}
+			case *schema.MavenConnection:
+				server.MavenSource, err = repos.NewMavenSource(e)
```
On line 196, you would want to include `extsvc.KindMaven`, otherwise this line of code would never be hit.

unknwon commented on cmd/gitserver/server/vcs_syncer.go:
```diff
@@ -302,3 +308,141 @@ func (s *PerforceDepotSyncer) RemoteShowCommand(ctx context.Context, remoteURL *
+// RemoteShowCommand returns the command to be executed for showing remote.
+func (s MavenArtifactSyncer) RemoteShowCommand(ctx context.Context, remoteURL *vcs.URL) (cmd *exec.Cmd, err error) {
+	// TODO: we return nil
+	return nil, nil
```
I'm pretty sure just do what `PerforceDepotSyncer` does is always safe as long as it is a Git repository:

```suggestion
	return exec.CommandContext(ctx, "git", "remote", "show", "./"), nil
```

unknwon commented on cmd/gitserver/server/server.go:
```diff
@@ -1325,20 +1325,22 @@ func (s *Server) cloneRepo(ctx context.Context, repo api.RepoName, opts *cloneOp
-		if cmd.Env == nil {
-			cmd.Env = os.Environ()
-		}
+		if cmd != nil {
```
Checking and relying on `nil` command is a bit weird, especially when the error is also `nil`. 

What about for MavenSyncer, just return a command that does something cheap and completely harmless like `git --version`?

olafurpg commented on internal/repos/maven.go:
```diff
@@ -0,0 +1,191 @@
+	}
+}
+
+func (s MavenSource) listAllRepos(ctx context.Context, results chan SourceResult) {
```
I don't think we should try to list all packages in a maven repository. There's no reliable way to list all packages for an arbitrary maven repository. The `coursier complete` command is only intended to power tab completions. We would end up creating millions of repos if we try to proxy all published artifacts on Maven Central.

It's better if we store the list of repos in a database similar to the schema in https://docs.google.com/document/d/1kq9UwAcDg8J43gTUtinnmLmJkqZw0WELKpb1JhFpCh0/edit#heading=h.fbsetjbel7u4

We would update that database when an uploaded LSIF dump adds a new `packageInformation` vertex.

shrouxm replied:
adding a db is an option, i've opted to store the info in the repos table using the opaque url fields

olafurpg commented on internal/extsvc/maven/coursier/coursier.go:
```diff
@@ -0,0 +1,65 @@
+
+// TODO: codeintel, we are manually constructing URLs to pass to 'cs get' rather than just calling
+// 'cs fetch --sources groupID:artifactID'. this is because roux couldn't figure out how to get
+// coursier to only fetch the sources we're interested rather than pulling down the entire
```
You can download a single `-sources.jar` with the following command
```
❯ cs fetch --intransitive com.google.guava:guava:30.1-jre --classifier sources
/Users/olafurpg/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/google/guava/guava/30.1-jre/guava-30.1-jre-sources.jar
```

olafurpg commented on internal/extsvc/maven/coursier/coursier.go:
```diff
@@ -0,0 +1,65 @@
+}
+
+func runCoursierCommand(ctx context.Context, repository string, args ...string) ([]string, error) {
+	cmd := exec.CommandContext(ctx, "coursier", append([]string{"-r", repository}, args...)...)
```
How do we ensure that `coursier` is always available in `$PATH`? On the JVM, I would  normally embed the `coursier` binary as a "resource" on the classpath. The slim launcher from https://git.io/coursier-cli is ~44kb and works on macos and linux

shrouxm replied:
this is a still a big TODO for the PR

olafurpg commented on internal/extsvc/maven/coursier/coursier.go:
```diff
@@ -0,0 +1,65 @@
+		urls = append(urls, url)
+	}
+
+	return runCoursierCommand(ctx, repository, append([]string{"get"}, urls...)...)
```
TIL about `coursier get`, never seen that before 😅

olafurpg commented on cmd/gitserver/server/vcs_syncer.go:
```diff
@@ -302,3 +308,141 @@ func (s *PerforceDepotSyncer) RemoteShowCommand(ctx context.Context, remoteURL *
+			return errors.Wrapf(err, "failed to git clean with output %q", string(output))
+		}
+
+		cmd = exec.CommandContext(ctx, "unzip", path, "-d", "./")
```
Not sure if this is needed, but might not hurt to always overwrite existing files
```suggestion
		cmd = exec.CommandContext(ctx, "unzip", "-o", path, "-d", "./")
```

shrouxm replied:
i think i'd actually rather not overwrite: if there's existing files, it means the git clean command failed, which means there could also be files that won't get overwritten but will erroneously be included in the commit

olafurpg commented on cmd/gitserver/server/vcs_syncer.go:
```diff
@@ -302,3 +308,141 @@ func (s *PerforceDepotSyncer) RemoteShowCommand(ctx context.Context, remoteURL *
+}
+
+func (s MavenArtifactSyncer) commitJars(ctx context.Context, dir GitDir, paths, versions []string) error {
+	for idx, path := range paths {
```
What do you think about using a separate repo for each package/version combination? I am concerned we will hit performance issues if we try to download all versions of a package.

olafurpg commented on cmd/gitserver/server/vcs_syncer.go:
```diff
@@ -302,3 +308,141 @@ func (s *PerforceDepotSyncer) RemoteShowCommand(ctx context.Context, remoteURL *
+			return errors.Wrapf(err, "failed to unzip with output %q", string(output))
+		}
+
+		// TODO: codeintel, need some extra logic here to set up maven repo properly for indexing
```
You need to generate an `lsif-java.json` file like here https://sourcegraph.com/maven/com.google.guava/guava/27.0.1-jre/-/blob/lsif-java.json

The logic in PackageHub to generate that file is implemented here, there's a different flow for maven packages and the JDK itself  https://github.com/sourcegraph/lsif-java/blob/ea78804200ebecdc04cca7a3d2a8b87b9446bc2a/packagehub/src/main/scala/com/sourcegraph/packagehub/PackageActor.scala#L349-L366

To determine the `"jvm"` field, we open a classfile inside the jar and read the first 8 bytes that give us the Java "major version". See https://github.com/sourcegraph/lsif-java/blob/ea78804200ebecdc04cca7a3d2a8b87b9446bc2a/lsif-semanticdb/src/main/java/com/sourcegraph/lsif_semanticdb/JavaVersion.java#L49-L65 To begin with, it's OK to hardcode `"jvm": 8` and we can implement support for other versions later

olafurpg reviewed (commented):
This is super exciting! I am looking forward to have built-in support for Maven repos on Sourcegraph 😄 

 The biggest open questions for me about the approach

- what do you think about using a database to list supported maven repos instead of using `coursier complete`?
- what do you think about using a separate repo for each package/version combination?

shrouxm commented:
i pushed a commit addressing all review comments but am running out of time in the day.

remaining work for the PR:
- test
- add the coursier binary to the relevant docker images

unknwon commented on cmd/gitserver/server/vcs_syncer.go:
```diff
@@ -302,3 +307,106 @@ func (s *PerforceDepotSyncer) RemoteShowCommand(ctx context.Context, remoteURL *
+		return nil
+	}
+
+	return errors.New(fmt.Sprintf("Maven repo %v not found", remoteURL))
```
```suggestion
	return errors.Errorf("Maven repo %q not found", remoteURL)
```

This is the shortcut 😁

PS, `%q` quotes the string, nit though.

unknwon commented on cmd/gitserver/server/vcs_syncer.go:
```diff
@@ -302,3 +307,106 @@ func (s *PerforceDepotSyncer) RemoteShowCommand(ctx context.Context, remoteURL *
+	return exec.CommandContext(ctx, "git", "--version"), s.commitJar(ctx, GitDir(tmpPath), groupID, artifactID, path, version)
+}
+
+var versionPattern = lazyregexp.New(`refs/heads/(.+)$`)
```
The regex looks more like a "headsPattern"? 🤔

unknwon commented on cmd/gitserver/server/vcs_syncer.go:
```diff
@@ -302,3 +307,106 @@ func (s *PerforceDepotSyncer) RemoteShowCommand(ctx context.Context, remoteURL *
+	return nil
+}
+
+type lsifJavaJson struct {
```
```suggestion
type lsifJavaJSON struct {
```

unknwon commented on internal/extsvc/types.go:
```diff
@@ -118,6 +119,9 @@ const (
 	// TypePhabricator is the (api.ExternalRepoSpec).ServiceType value for Phabricator projects.
 	TypePhabricator = "phabricator"
 
+	// TypePhabricator is the (api.ExternalRepoSpec).ServiceType value for Phabricator projects.
```
This comment doesn't seem right?

efritz replied:
```suggestion
// TypeMaven is the (api.ExternalRepoSpec).ServiceType value for Maven artifacts.
```

unknwon commented:
> @unknwon still need to test everything, but wanted to ask for a review before i dive into that just to make sure the direction i'm going seems reasonable to you.

The direction looks good to me 👍 Please LMK when it's ready for a formal review.

efritz commented on internal/conf/reposource/maven.go:
```diff
@@ -0,0 +1,22 @@
+
+func DecomposeMavenPath(path string) (groupID, artifactID, version string) {
+	split := strings.Split(strings.TrimPrefix(path, "/"), "/")
+	return split[0], split[1], split[2]
```
Should check indexes here so we don't crash. Return an error from this method?

efritz commented on internal/extsvc/maven/coursier/coursier.go:
```diff
@@ -0,0 +1,67 @@
+}
+
+func runCoursierCommand(ctx context.Context, config *schema.MavenConnection, args ...string) ([]string, error) {
+	cmd := exec.CommandContext(ctx, "coursier", args...)
```
Need to ensure this binary is in the container?

efritz commented on internal/repos/maven.go:
```diff
@@ -0,0 +1,168 @@
+		artifactID := split[1]
+		versions, err := coursier.ListVersions(ctx, s.config, groupID, artifactID)
+		if err != nil {
+			results <- SourceResult{Err: err}
```
I this assumed to be buffered, or should you read from ctx.Done() as well? Can this block forever?

olafurpg commented:
I'm working on a new PR that builds on top of this one. Closing this PR for now.
